Allocate identity hash on demand rather than at object allocation time

blairmcg opened the pull request:
This improves identity hash distribution when allocating large numbers of composite objects.

bernhardkohlhaas commented:
Hi @blairmcg ,

I have a question, since I was also playing around with the lazy hash init. Since this is my first time looking at the VM, I didn't get all the way, but I noticed something peculiar and I'm wondering, if you have an explanation:
- For test purposes I modified ObjectMemory::nextIdentityHash() in ObjMemPriv.inl to exclude a particular hash value (0xFFF0) from the list of values it sets to m_idHash. (I was planning to use that as a "Not Initialized" cookie).
- Then I made a change to primitive IdentityHash16 in primasm.asm to test for that value and - if matched - trigger a break point in the debugger.

And I indeed found that every once in a while the break point would be hit, even though there shouldn't really have been an object with that hash. I'm wondering now where those objects were coming from. Perhaps you have an explanation.

In case you'd like to try this yourself, my changes are at https://github.com/bernhardkohlhaas/DolphinVM/commit/87b42ee13f09fdbcaaa20c754120f852248aa940  in https://github.com/bernhardkohlhaas/DolphinVM/tree/HashLazyInit

(This was just code for testing and playing around, so there is a lot of junk in there, so please disregard some of the mess)

The way I reproduced the breakpoint was:
- Set a breakpoint in primasm.asm, line 1985, 
- Boot fresh image with the VM
- Launch the image and attach the debugger to the running Dolphin7.exe process
- Then in the image open a new workspace and in there type "Object new hash" (without quotation marks).

Usually after hitting the <space> key after "new" the breakpoint would trigger.

Could you please explain, why there were still objects with that hash (although I worked from an image that was booted with my modified VM)?

Thanks so much in advance for your help,
Bernhard

blairmcg commented:
@bernhardkohlhaas , there is nothing fundamentally wrong with what you have done, you are just seeing some objects that have survived from the boot image (dboot.img7).
I wondered about the same issue, and having decided that zero was the obvious choice for an uninitialized identity hash, I added a bit of code to the VM to dump out the objects in the dboot image with zero id hash. There turned out to only be three, and none of them stored in any hashed collection, which was lucky. Since the boot image can't be rebuilt at all easily, the luck will persist so I stopped worrying about patching it up.
BTW: I moved the primitive into C++ from assembler, taking advantage of some other work I'd done to reduce the cost of calling C++ primitives. This will make it easier to experiment with other RNGs, although I'm not sure that there would be much to gain given that only 16-bits of the result can be used.

bernhardkohlhaas commented:
@blairmcg , thanks so much for the explanation. Yes, it's very nice to have the primitive in C++ now, though tinkering with assembler was exciting in its own way. :)
